lib/Applet.pm

```diff
@@ -741,14 +742,19 @@ sub insertObject {
 		foreach my $key (keys %param_hash) {
 			$javaParameters .= qq!<param name="$key" value="$param_hash{$key}">\n!;
 			$flashParameters .= uri_escape($key) . '=' . uri_escape($param_hash{$key}) . '&';
-			$webgeogebraParameters .= qq!data-param-$key = "$param_hash{$key}"\n!;
 		}
 		$flashParameters =~ s/\&$//; # trim last &
-		$webgeogebraParameters = qq!<article class="geogebraweb"
-			data-param-id     = "$appletName"
-			data-param-width  = "$width"
-			data-param-height = "$height"
-			!. $webgeogebraParameters . qq!\n></article>!;
+		$param_hash{id}     = $appletName;
+		$param_hash{height} = $height;
+		$param_hash{width}  = $width;
+		$webgeogebraParameters .= encode_json(\%param_hash) . ';';
+		$webgeogebraParameters .= qq!
+			var $appletName = new GGBApplet(params, true);
+			window.addEventListener('load', function() {
+				$appletName.setHTML5Codebase('https://geogebra.org/apps/latest/web3d/');
+				$appletName.inject('$appletName');
+			});
+		</script>!;
```

Comment:
I don't like the global variable that is created.  Nor do I like the inline javascript.  This could certainly be done in the ww_applet_support.js script instead, and only pass data attributes like before.

To do this, change the `article` to a `div`, add the `id` (as you have above), and remove `param-` from the added data attributes.   Then add the following to line 627 of `webwork2/htdocs/js/apps/AppletSupport/ww_applet_support.js`:
```javascript
if (ww_applet_list[appletName].type == 'geogebraweb') {
	var container = document.getElementById(appletName);
	var ggbApplet = new GGBApplet(Object.assign({}, container.dataset), true);
	ggbApplet.setHTML5Codebase('https://geogebra.org/apps/latest/web3d/');
	ggbApplet.inject(appletName);
}
```

The attached zip file contains three patches that demonstrate all of this.  The cleaner-pg-2.16 patch applies to the PG-2.16 branch (instead of this pull request), and is included to demonstrate directly what is changed from there.  The cleaner-update-geogebra-api.patch applies to this pull request branch and does the same thing.  The webwork2-update-geogebra-applet-js.patch applies to the WeBWorK-2.16 branch of the webwork2 code and makes the suggested javascript changes.

[patches.zip](https://github.com/openwebwork/pg/files/6670440/patches.zip)

Really, what needs to be done is for the ww_applet_support.js script to be completely rewritten and remove the flash and java support, and clean it up to the barebones javascript needed for Geogebra applets with no globals.

Reply:
That's a really neat trick with `dataset`!

Reply:
I guess it is a bit silly to complain about inline javascript and global variables here, when a few lines below in Applets.pm is another instance of inline javascript and a global variable.

Reply:
it seems that even with this strategy, the applet object is still available as a global. Not sure how that's happening...